go/parquet/metadata/statistics.go

```diff
@@ -0,0 +1,517 @@
+// they are being populated
+type TypedStatistics interface {
+	// Type is the underlying physical type for this stat block
+	Type() parquet.Type
+	// Returns true if there is a min and max value set for this stat object
+	HasMinMax() bool
+	// Returns true if a nullcount has been set
+	HasNullCount() bool
+	// returns true only if a distinct count has been set
+	// current implementation does of the writer does not automatically populate
+	// the distinct count right now.
+	HasDistinctCount() bool
+	NullCount() int64
+	DistinctCount() int64
+	NumValues() int64
+	// return the column descriptor that this stat object was initialized with
+	Descr() *schema.Column
+
+	// Encode the current min value and return the bytes. ByteArray does not
+	// include the len in the encoded bytes, otherwise this is identical to
+	// plain encoding
+	EncodeMin() []byte
+	// Encode the current max value and return the bytes. ByteArray does not
+	// include the len in the encoded bytes, otherwise this is identical to
+	// plain encoding
+	EncodeMax() []byte
+	// Populate an EncodedStatistics object from the current stats
+	Encode() (EncodedStatistics, error)
+	// Resets all values to 0 to enable reusing this stat object for multiple
+	// columns, by calling Encode to get the finished values and then calling
+	// reset
+	Reset()
+	// Merge the min/max/nullcounts and distinct count from the passed stat object
+	// into this one.
+	Merge(TypedStatistics)
+}
+
+type statistics struct {
+	descr            *schema.Column
+	hasMinMax        bool
+	hasNullCount     bool
+	hasDistinctCount bool
+	mem              memory.Allocator
+	nvalues          int64
+	stats            EncodedStatistics
+	order            schema.SortOrder
+
+	encoder encoding.TypedEncoder
+}
+
+func (s *statistics) incNulls(n int64) {
+	s.stats.NullCount += n
+	s.hasNullCount = true
+}
+func (s *statistics) incDistinct(n int64) {
+	s.stats.DistinctCount += n
+	s.hasDistinctCount = true
+}
+
+func (s *statistics) Descr() *schema.Column  { return s.descr }
+func (s *statistics) Type() parquet.Type     { return s.descr.PhysicalType() }
+func (s *statistics) HasDistinctCount() bool { return s.hasDistinctCount }
+func (s *statistics) HasMinMax() bool        { return s.hasMinMax }
+func (s *statistics) HasNullCount() bool     { return s.hasNullCount }
+func (s *statistics) NullCount() int64       { return s.stats.NullCount }
+func (s *statistics) DistinctCount() int64   { return s.stats.DistinctCount }
+func (s *statistics) NumValues() int64       { return s.nvalues }
+
+func (s *statistics) Reset() {
+	s.stats.NullCount = 0
+	s.stats.DistinctCount = 0
+	s.nvalues = 0
+	s.hasMinMax = false
+	s.hasDistinctCount = false
+	s.hasNullCount = false
+}
+
+func (s *statistics) merge(other TypedStatistics) {
+	s.nvalues += other.NumValues()
+	if other.HasNullCount() {
+		s.stats.NullCount += other.NullCount()
+	}
+	if other.HasDistinctCount() {
+		s.stats.DistinctCount += other.DistinctCount()
```

Comment:
I'm pretty sure this is incorrect?  distinct counts would need to retain set information and combined via set to get total distinct count.

Reply:
Funny enough this is also how the C++ implementation works too. Which i believe doesn't currently set distinct count to anything yet anyways.

Reply:
yeah, I would think maybe this should be ignored?  or at least a comment acknowledging it is incorrect.

Reply:
Sure, i'll add a comment acknowledging it

Reply:
comment added.